test/cpp/end2end/async_end2end_test.cc

```diff
@@ -67,10 +67,10 @@ namespace {
 void* tag(int i) { return (void*)(intptr_t)i; }
 
 #ifdef GPR_POSIX_SOCKET
-static int assert_non_blocking_poll(struct pollfd* pfds, nfds_t nfds,
+static int non_blocking_poll(struct pollfd* pfds, nfds_t nfds,
                                     int timeout) {
-  GPR_ASSERT(timeout == 0);
-  return poll(pfds, nfds, timeout);
+  /* ignore timeout and always use timeout 0 */
```

Comment:
Can I suggest the following then:

a thread local 'g_is_async_end2end_test_thread' that is true for any thread spawned by the end2end test, and false otherwise.

We can test that value and assert the timeout is zero iff that thread local is true.

This is the only test that prevents #2228 from reappearing.